File: ecmascript/bob/bob.js
const isShout = message => message.match(/[A-Z]/) && message.toUpperCase() === message;
const isQuestion = message => message.endsWith('?');
const isForceful = message => message.endsWith('!');
const isForcefulQuestion = message => isQuestion(message) && isShout(message);
const isSilence = message => message.match(/^\s*$/) !== null;

class Bob {
  hey(message) {
    const testResult = tests.find(test => test.predicate(message));
    return testResult ? testResult.response : Bob.responses.whatever;
  }
}

Bob.responses = {
  whatever: 'Whatever.',
  whoa: 'Whoa, chill out!',
  sure: 'Sure.',
  calmDown: 'Calm down, I know what I\'m doing!',
  fine: 'Fine. Be that way!'
};

const tests = [
  {predicate: isSilence, response: Bob.responses.fine},
  {predicate: isForcefulQuestion, response: Bob.responses.calmDown},
  {predicate: isQuestion, response: Bob.responses.sure},
  {predicate: isShout, response: Bob.responses.whoa},
  {predicate: isForceful, response: Bob.responses.whatever},
];

export default Bob;

File: go/bob/bob.go
package bob

import "regexp"

var shouting = regexp.MustCompile(`^[^a-z]+\!?$`)
var shoutingGibberish = regexp.MustCompile(`^[A-Z]+$`)
var question = regexp.MustCompile(`\?$`)
var forcefulQuestion = regexp.MustCompile(`^[A-Z\s]+\?$`)
var noLetters = regexp.MustCompile(`^[^a-zA-Z]+$`)
var silence = regexp.MustCompile(`^\s*$`)
var endWithWhitespace = regexp.MustCompile(`\s$`)
var questionEndWithWhitespace = regexp.MustCompile(`\?\s+$`)
var nonQuestionEndWithWhitespace = regexp.MustCompile(`[^\?]*\s$`)

const chillOut = `Whoa, chill out!`
const sure = `Sure.`
const whatever = `Whatever.`
const calmDown = `Calm down, I know what I'm doing!`
const beThatWay = `Fine. Be that way!`

func Hey(remark string) string {
	switch {
	case silence.MatchString(remark):
		return beThatWay
	case questionEndWithWhitespace.MatchString(remark):
		return sure
	case nonQuestionEndWithWhitespace.MatchString(remark):
		return whatever
	case endWithWhitespace.MatchString(remark):
		return sure
	case forcefulQuestion.MatchString(remark):
		return calmDown
	case question.MatchString(remark):
		return sure
	case noLetters.MatchString(remark):
		return whatever
	case shouting.MatchString(remark):
		return chillOut
	case shoutingGibberish.MatchString(remark):
		return chillOut
	default:
		return whatever
	}
}
